### PR TITLE
add structure for record updates

### PR DESCRIPTION
**Pull request by willscott**
this builds the top level scaffold expected to be consumed by reputation systems

* A new command for now to indicate all work by a given bot is done, at which point all finalized tasks from that bot are collected into a record update that is set as the new 'head'
* A graphql endpoint to query against this chain.

### CODE REVIEW

File: controller/state/record_update.go [hunk 1]
```diff
@@ -0,0 +1,10 @@
+package state
+
+type record_update_status int
+
+const (
+	LATEST_UPDATE     record_update_status = 1
+	PREVIOUS_UPDATE   record_update_status = 2
+	UNATTACHED_RECORD record_update_status = 10
+	ATTACHED_RECORD   record_update_status = 11
```
Comment (mvdan) [inline controller/state/record_update.go:9]: interesting choice of casing, for Go :)

File: controller/state/statedb.go [hunk 2]
```diff
@@ -348,10 +348,18 @@ func (s *stateDB) Update(ctx context.Context, taskID string, req tasks.UpdateTas
...
+			finalized, err := updatedTask.Finalize(ctx, txContextStorer(ctx, tx))
 			if err != nil {
 				return err
 			}
+			flink, err := linkProto.Build(ctx, ipld.LinkContext{}, finalized, txContextStorer(ctx, tx))
+			if err != nil {
+				return err
+			}
+
+			if _, err = tx.ExecContext(ctx, addHeadSQL, flink.(linksystem.Link).Cid.String(), time.Now(), req.WorkedBy.String(), UNATTACHED_RECORD); err != nil {
```
Comment (mvdan) [inline controller/state/statedb.go:360]: might as well use `:=` here

Comment (mvdan) [reply controller/state/statedb.go:360]: and perhaps split the line as it's quite long

File: controller/state/statedb.go [hunk 3]
```diff
@@ -483,6 +501,109 @@ func (s *stateDB) countTasks(ctx context.Context) (int, error) {
...
+	}
+
+	rcrds := []tasks.AuthenticatedRecord{}
+
+	for itms.Next() {
+		var lnk string
+		if err := itms.Scan(&lnk); err != nil {
+			return err
+		}
+		if _, err = tx.ExecContext(ctx, updateHeadSQL, ATTACHED_RECORD, lnk); err != nil {
```
Comment (mvdan) [inline controller/state/statedb.go:573]: ditto `:=`

File: controller/state/statedb.go [hunk 4]
```diff
@@ -483,6 +501,109 @@ func (s *stateDB) countTasks(ctx context.Context) (int, error) {
...
+			return err
+		}
+		c, err := cid.Decode(lnk)
+		if err != nil {
+			return err
+		}
+		rcrd := tasks.Type.AuthenticatedRecord.Of(c, sig)
+		rcrds = append(rcrds, rcrd)
+	}
+	rcrdlst := tasks.Type.List_AuthenticatedRecord.Of(rcrds)
```
Comment (mvdan) [inline controller/state/statedb.go:587]: reading these names is making my head hurt a little... how about just `records` and `recordList`?

File: tasks/gen.go [hunk 5]
```diff
@@ -138,6 +139,18 @@ func main() {
...
+	// reputation api
+	ts.Accumulate(schema.SpawnStruct("AuthenticatedRecord", []schema.StructField{
+		schema.SpawnStructField("Record", "Link_FinishedTask", false, false),
+		schema.SpawnStructField("Signature", "Bytes", false, false),
+	}, schema.SpawnStructRepresentationMap(map[string]string{})))
+	ts.Accumulate(schema.SpawnList("List_AuthenticatedRecord", "AuthenticatedRecord", false))
+	ts.Accumulate(schema.SpawnStruct("RecordUpdate", []schema.StructField{
+		schema.SpawnStructField("Records", "List_AuthenticatedRecord", false, false),
+		schema.SpawnStructField("SigPrev", "Bytes", false, false),
+		schema.SpawnStructField("Previous", "Link", false, false),
```
Comment (mvdan) [inline tasks/gen.go:151]: Link to...? I imagine this could/shoudl be something like Link_RecordUpdate.

**Review by mvdan — APPROVED** [review]
Maybe document somewhere why you're using a chain-like structure with signatures and a "head". I can guess why we'd want that, but it's not entirely obvious.

File: controller/state/record_update.go (hunk 1 above)
Comment (willscott) [reply controller/state/record_update.go:9]: thoughts on preferred case for these constants? jut `LatestUpdate`?

File: controller/state/statedb.go (hunk 4 above)
Comment (willscott) [reply controller/state/statedb.go:587]: in the reputation system parlance we call `Task` a `Record`, and then have a surrounding `AuthenticatedRecord` that is a signature over the cid of the record.

File: tasks/gen.go (hunk 5 above)
Comment (willscott) [reply tasks/gen.go:151]: this seems to be a bug in graphql code gen. If i have a cycle of links (struct includes link to itself) we end up with a circular definition.

**Pull request merged**